### DISCUSSION TITLE
Example kotlin scripting with datasettes

### DISCUSSION THREAD

**Discussion by yschimke** [simonw/datasette#924] [General]
Hopefully useful sample code if you want to extract data out of a datasette and do something with it in kotlin.  The script should run just with kotlin and nothing else installed.

https://github.com/yschimke/okurl-scripts/blob/master/commands/sydney-toilets.main.kts

Results in iTerm

<img width="665" alt="image" src="https://user-images.githubusercontent.com/231923/89938583-9bd06480-dc0e-11ea-9530-9cd000aa55fd.png">

**Comment by simonw** [comment]
I love this example!

I just updated https://australian-dunnies.now.sh/ to run on the latest version of Vercel since it was previously running on Now v1 which is in the process of shutting down this week. Let me know if anything breaks!

**Reply by yschimke** [reply]
I've just updated for the API changes.

**Reply by yschimke** [reply]
Can you do me a favour and download that script and try running it?  You can comment out the `show(staticMap` statement and just see the other results.

**Reply by simonw** [reply]
I tried but I don't know anything about the Kotlin ecosystem so I'm not sure what I'm doing. Here's where I got to...
```
~ % brew install kotlin
==> Downloading https://github.com/JetBrains/kotlin/releases/download/v1.3.72/ko
==> Downloading from https://github-production-release-asset-2e65be.s3.amazonaws
######################################################################## 100.0%
🍺  /usr/local/Cellar/kotlin/1.3.72: 118 files, 58.4MB, built in 4 seconds
~ % cd /tmp
/tmp % wget 'https://raw.githubusercontent.com/yschimke/okurl-scripts/master/commands/sydney-toilets.main.kts'
/tmp % kotlin sydney-toilets.main.kts 
No Java runtime present, requesting install.
/tmp % brew cask install java
==> Downloading https://download.java.net/java/GA/jdk14.0.2/205943a0976c4ed48cb16f1043c5c647/12/GPL/openjdk-1
######################################################################## 100.0%
...
🍺  java was successfully installed!
/tmp % kotlin sydney-toilets.main.kts
:: problems summary ::
:::: WARNINGS
		module not found: com.github.yschimke#okurl-script;1.0.0

	==== central: tried

	  https://repo1.maven.org/maven2/com/github/yschimke/okurl-script/1.0.0/okurl-script-1.0.0.pom

	  -- artifact com.github.yschimke#okurl-script;1.0.0!okurl-script.jar:

	  https://repo1.maven.org/maven2/com/github/yschimke/okurl-script/1.0.0/okurl-script-1.0.0.jar

error: cannot access script base class 'org.jetbrains.kotlin.mainKts.MainKtsScript'. Check your module classpath for missing or conflicting dependencies (sydney-toilets.main.kts:3:1)
error: unresolved reference: Repository (sydney-toilets.main.kts:3:7)
error: unresolved reference: DependsOn (sydney-toilets.main.kts:4:7)
error: unresolved reference: CompilerOptions (sydney-toilets.main.kts:5:7)
...
```

**Reply by yschimke** [reply]
I've just pushed hopefully a fix.  The scripts (latest commit) should run like bash scripts if you have kotlinc-jvm on your path.

**Reply by yschimke** [reply]
![image](https://user-images.githubusercontent.com/231923/90066043-15d31d00-dce5-11ea-80d8-c12e6cc3c084.png)

**Reply by yschimke** [reply]
But they also run inside Intellij by right clicking and running.  And have type safe editing there also.

**Reply by yschimke** [reply]
![image](https://user-images.githubusercontent.com/231923/90066218-5c287c00-dce5-11ea-8821-5971f2ba6542.png)

**Reply by simonw** [reply]
I tried again with this:
```
/tmp % ./sydney-toilets.main.kts  
OpenJDK 64-Bit Server VM warning: Options -Xverify:none and -noverify were deprecated in JDK 13 and will likely be removed in a future release.
:: problems summary ::
:::: WARNINGS
		module not found: com.github.yschimke#okurl-script;1.0.0

	==== central: tried

	  https://repo1.maven.org/maven2/com/github/yschimke/okurl-script/1.0.0/okurl-script-1.0.0.pom

	  -- artifact com.github.yschimke#okurl-script;1.0.0!okurl-script.jar:

	  https://repo1.maven.org/maven2/com/github/yschimke/okurl-script/1.0.0/okurl-script-1.0.0.jar

error: cannot access script base class 'org.jetbrains.kotlin.mainKts.MainKtsScript'. Check your module classpath for missing or conflicting dependencies (sydney-toilets.main.kts:3:1)
```
It looks like I need to install dependencies using Maven, but I don't know how to do that on this machine (it didn't have a JDK installed until half an hour ago!)

**Reply by yschimke** [reply]
You don't, I'll look into it before I waste any more of your time.  Kotlin has features to download the jars for you.
